dts: pwm : flag to enable a complementary output of a pwm channel pin

FRASTM opened the pull request:
Depending on the mcus and the timer instance, several channels
can enable the complementary output for the PWM signal
This flag completes the PWM_POLARITY for a PWM channel.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

gmarull commented on include/dt-bindings/pwm/pwm.h:
```diff
@@ -24,4 +24,17 @@
+ * The `PWM_COMPLEMENTARY` flag is used with pwm_pin_set()
+ * to enable the complementary output of the timer channel.
+ * At the moment, only used by stm32 devices where some channels
+ * of some Timers have that capability.
```
```suggestion
 * This flag can be used with any of the `pwm_pin_set_*` API calls to indicate
 * that the PWM signal has to be routed to the complementary output channel.
 * This feature is only available on certain SoC families, refer to the
 * binding's documentation for more details.
```

gmarull commented on include/dt-bindings/pwm/pwm.h:
```diff
@@ -24,4 +24,19 @@
 
+/**
+ * @name PWM complementary flag for output pins
+ * The `PWM_COMPLEMENTARY` flag is used with pwm_pin_set()
+ * to enable the complementary output of some channels of some timers.
```
```suggestion
 * This flag can be used with any of the `pwm_pin_set_*` API calls to indicate
 * that the PWM signal has to be routed to the complementary output channel.
 * This feature is only available on certain SoC families, refer to the
 * binding's documentation for more details.
```

FRASTM replied:
ok

henrikbrixandersen commented:
As this is rather SoC specific (and not generically PWM related), have you considered if this could be configured using devicetree instead? Isn't this for routing of the generated PWM signal?

gmarull commented:
> As this is rather SoC specific (and not generically PWM related), have you considered if this could be configured using devicetree instead? Isn't this for routing of the generated PWM signal?

For reference, there's another proposal I made that is vendor-specific: https://github.com/zephyrproject-rtos/zephyr/pull/42119#issuecomment-1021226073

henrikbrixandersen reviewed (changes requested):
Blocking for now as I think this is too SoC specific to be put into the generic PWM API. I would like to have more discussion about this (e.g. in the API meeting on Tuesdays) first.

carlescufi commented:
API meeting:

- @henrikbrixandersen proposes using a similar approach to the one suggested [here](https://github.com/zephyrproject-rtos/zephyr/pull/39767#issuecomment-1032798773)

henrikbrixandersen commented:
@FRASTM Do you plan on reworking this with an approach similar as to that for GPIO in https://github.com/zephyrproject-rtos/zephyr/pull/39767?

FRASTM commented:
> @FRASTM Do you plan on reworking this with an approach similar as to that for GPIO in #39767?

Yes, I guess this means : abandon this one and updating the https://github.com/zephyrproject-rtos/zephyr/pull/42119
with a `PWM_STM32_COMPLEMENTARY` flag in a  include/dt-bindings/pwm/stm32_pwm.h

FRASTM committed:
include: drivers: pwm: extend the pwm flags to be soc specific

Add 8 bits to the pwm_flags_t so that upper byte is reserved
for non-standard but soc specific dts flags and keeping the low
Byte for standard flags.
Some of SoC like STM32 mcus can then define their own flags
in their dt-bindings/pwm.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
reabse on 44fdeb1dd89dc6c0ba9cd0bb2cb936f166fb8560

FRASTM commented:
closing after the merge of https://github.com/zephyrproject-rtos/zephyr/pull/42119/
